src/Pact/Types/PactValue.hs

```diff
@@ -90,3 +106,211 @@ toPactValueLenient t = case toPactValue t of
   Right (PLiteral (LInteger l)) -> PLiteral (LDecimal (fromIntegral l))
   Right v -> v
   Left _ -> PLiteral $ LString $ renderCompactText t
+
+
+
+-- |  Estimate of number of bytes needed to represent data type
+--
+-- Assumptions: GHC, 64-bit machine
+-- General approach:
+--   Memory Consumption = Constructor Header Size + Cost of Constructor Field(s)
+--   Cost of Constructor Field(s)* = 1 word per field + cost of each field's value
+-- (*) See Resource 2 for exceptions to these rules
+-- Resources:
+-- 1. http://wiki.haskell.org/GHC/Memory_Footprint
+-- 2. https://stackoverflow.com/questions/3254758/memory-footprint-of-haskell-data-types
+
+class SizeOf t where
+  sizeOf :: t -> Bytes
+
+
+type Bytes = Int64
+
+-- TOOD Figure out how to detect machine arch
+-- | "word" is 4 bytes on 32-bit arch, but 8 bytes on 64-bit
+wordSize32, wordSize64, wordSize :: Bytes
+wordSize32 = 4
+wordSize64 = 8
+wordSize = wordSize64
+
+-- | Constructor header is 1 word
+headerCost :: Bytes
+headerCost = 1 * wordSize
+
+-- | In general, each constructor field costs 1 word
+constructorFieldCost :: Int64 -> Bytes
+constructorFieldCost numFields = numFields * wordSize
+
+-- | Total cost for constructor
+constructorCost :: Int64 -> Bytes
+constructorCost numFields = headerCost + (constructorFieldCost numFields)
+
+
+instance SizeOf PactValue where
+  sizeOf (PLiteral l) = (constructorCost 1) + (sizeOf l)
+  sizeOf (PList v) = (constructorCost 1) + (sizeOf v)
+  sizeOf (PObject o) = (constructorCost 1) + (sizeOf o)
+  sizeOf (PGuard g) = (constructorCost 1) + (sizeOf g)
+
+instance SizeOf Literal where
+  sizeOf (LString t) = (constructorCost 1) + (sizeOf t)
+  sizeOf (LInteger i) = (constructorCost 1) + (sizeOf i)
+  sizeOf (LDecimal d) = (constructorCost 1) + (sizeOf d)
+  sizeOf (LBool _) = (constructorCost 1) + 0
```

Comment:
You only pay a pointer cost for using Bool -- all instances of True are pointers to the same object. You can prove this to yourself --

```
λ ghci
GHCi, version 8.6.4: http://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /home/greg/git/kadena/chainweb-node/.ghci
Prelude> :set -XMagicHash
Prelude> import GHC.Prim
Prelude GHC.Prim> import GHC.Int
Prelude GHC.Prim GHC.Int> let a = True
Prelude GHC.Prim GHC.Int> let b = True
Prelude GHC.Prim GHC.Int> I# (reallyUnsafePtrEquality# a b)
1
```

I think this may actually be true for all data types having only nullary constructors, but I'm not sure.

Reply:
@gregorycollins this is about storage, not RAM. In general we're not charging for RAM, assuming that storage and CPU dominates.

Reply:
Therefore I think the current sizing is fine, maybe even too small because of JSON storage, but let's not go there

Reply:
@slpopejoy @gregorycollins I'll be resolving this comment.